### DISCUSSION TITLE
How to serialize classes (not instances)?

### DISCUSSION THREAD

**Discussion by rdbisme** [pydantic/pydantic#9799] [Question] [answered]
Hello, I have something like this: 

```python
class MyType(BaseModel): 
   actions: set[type[Action]]
```

and I'd like to serialize `type[Action]` to it's fqn string (`"package.module.MyAction"`). How do I do this? I tried to define a metaclass `SerializeableMeta` for `Action`, where I defined `__get_pydantic_core_schema__`. But it doesn't seem to work. Still getting `Unable to serialize unknown type 'SerializeableMeta'.

**Comment by Viicos** [comment, marked as answer]
You can use a [`PlainSerializer`](https://docs.pydantic.dev/dev/api/functional_serializers/#pydantic.functional_serializers.PlainSerializer):

```python
class Action:
    pass


def serialize_to_fqn(value: type[Action]) -> str:
    return f"{value.__module__}.{value.__qualname__}"


class MyType(BaseModel): 
   actions: set[Annotated[type[Action], PlainSerializer(serialize_to_fqn)]]

MyType(actions={Action}).model_dump()
#> {'actions': {'__main__.Action'}}
```

**Reply by rdbisme** [reply]
Thanks a lot. That's what I ended up using! 
Any chance I might abuse your kindness to have one version which implements this with `__get_pydantic_core_schema__`. It's not totally clear how to use that to me.

**Reply by Viicos** [reply]
Well if you look at the definition of `PlainSerializer`, it uses `__get_pydantic_core_schema__` under the hood :)

**Reply by Helveg** [reply]
I'd like to do the same, but I'd love to define this logic on the class itself, instead of having to annotate it when it's used in a type hint. Is there any way to define this logic on the `Action` class itself here?

**Reply by Viicos** [reply]
That's currently not possible but is a request that we get a lot, but you can check some of the issues labeled as https://github.com/pydantic/pydantic/labels/topic-types%20support (both in issues and discussions). We have some ongoing work for a new API design but it is challenging.
